### PR TITLE
Fixes #790 - new template cisco_nxos_show_environment.textfsm 

### LINKED ISSUE #790: cisco_nxos_show_environment return an empty list
##### ISSUE TYPE
 - Bug Report

##### TEMPLATE USING

```
Value POWER_SUPPLY (\d+)
Value POWER_SUPPLY_MODEL (\S+)
Value POWER_SUPPLY_OUTPUT (\d+)
Value POWER_SUPPLY_CAPACITY (\d+)
Value POWER_SUPPLY_STATUS (\w+)
Value FAN (\S+)
Value FAN_STATUS (\S+)
Value TEMPERATURE_MODULE (\d+)
Value TEMPERATURE_SENSOR (\S+)
Value TEMPERATURE_MAJOR_THRESH (\d+)
Value TEMPERATURE_MINOR_THRESH (\d+)
Value TEMPERATURE_CURRENT (\d+)
Value TEMPERATURE_STATUS (\S+)

Start
  ^Power\s+Actual\s+Total -> Power
  ^Fan\s+Model\s+Hw\s+Status -> Fan

Power
  ^${POWER_SUPPLY}\s+${POWER_SUPPLY_MODEL}\s+${POWER_SUPPLY_OUTPUT}\s+\w+\s+${POWER_SUPPLY_CAPACITY}\s+\w+\s+${POWER_SUPPLY_STATUS} -> Record
  ^Module\s+Model\s+Draw\s+Allocated\s+Status -> Start

Fan
  ^Fan\s+Zone\s+Speed -> Temperature
  ^${FAN}\s+\S+\s+\S+\s+${FAN_STATUS} -> Record
  ^Temperature: -> Temperature

Temperature
  ^${TEMPERATURE_MODULE}\s+${TEMPERATURE_SENSOR}(\s+)?\(\S+\)\s+${TEMPERATURE_MAJOR_THRESH}\s+${TEMPERATURE_MINOR_THRESH}\s+${TEMPERATURE_CURRENT}\s+${TEMPERATURE_STATUS} -> Record
```

##### SAMPLE COMMAND OUTPUT

```
Power Supply:
Voltage: 12 Volts
Power                             Actual      Actual        Total
Supply    Model                   Output      Input      Capacity    Status
                                (Watts )    (Watts )     (Watts )
-------  -------------------  ----------  ----------  ----------  --------------
1        N9K-PAC-650W              101 W       103 W       650 W     Ok
2        N9K-PAC-650W               96 W       102 W       650 W     Ok


Power Usage Summary:
--------------------
Power Supply redundancy mode (configured)                PS-Redundant
Power Supply redundancy mode (operational)               PS-Redundant

Total Power Capacity (based on configured mode)             650.00 W
Total Grid-A (first half of PS slots) Power Capacity        650.00 W
Total Grid-B (second half of PS slots) Power Capacity       650.00 W
Total Power of all Inputs (cumulative)                     1300.00 W
Total Power Output (actual draw)                            197.00 W
Total Power Input (actual draw)                             206.00 W
Total Power Allocated (budget)                                N/A
Total Power Available for additional modules                  N/A

Fan:
---------------------------------------------------------------------------
Fan             Model                Hw     Direction       Status
---------------------------------------------------------------------------
Fan1(sys_fan1)  N9K-C9300-FAN2       --     front-to-back   Ok
Fan2(sys_fan2)  N9K-C9300-FAN2       --     front-to-back   Ok
Fan3(sys_fan3)  N9K-C9300-FAN2       --     front-to-back   Ok
Fan_in_PS1      --                   --     front-to-back   Ok
Fan_in_PS2      --                   --     front-to-back   Ok
Fan Zone Speed: Zone 1: 0x80
Fan Air Filter : NotSupported


Temperature:
--------------------------------------------------------------------
Module   Sensor        MajorThresh   MinorThres   CurTemp     Status
                       (Celsius)     (Celsius)    (Celsius)
--------------------------------------------------------------------
1        FRONT           70              42          24         Ok
1        BACK            80              70          26         Ok
1        CPU             90              80          42         Ok
1        TD2-1           105             90          35         Ok
1        NS-1            105             90          37         Ok

```

##### SUMMARY
cisco nxos show environment returns empty[]. Show environment temperature works perfectly. 

##### STEPS TO REPRODUCE

```
	swenv = sshsw1.send_command("show environment")
	postntcenv = parse_output(platform=cisco_nxos, command="show environment", data=swenv)
        print (postntcenv)
```

##### ACTUAL RESULTS
```
   []
```

### PR DESCRIPTION
**Pull request by diepes**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT

cisco_nxos
cisco_nxos_show_environments.textfsm (Current broken)
cisco_nxos_show_environment.textfsm  (New)

##### SUMMARY

Fixes #790 
Current template "cisco_nxos_show_environments.textfsm" not fixable, notice extra "s" and it tries to extract two tables, Power and Fan from output.
Create new template "cisco_nxos_show_environment.textfsm" that handle old .raw and new #790 output, but only extracts the power table.

Will request from user output of more specific commands to be added as additional templates.
e.g.  show environment fan

See: tests/cisco_nxos/show_environments/cisco_nxos_show_environments.yml
vs:   tests/cisco_nxos/show_environment/cisco_nxos_show_environment.yml
Same .raw parsed.
```

```

### CODE REVIEW

File: templates/index [hunk 1]
```diff
@@ -276,9 +276,10 @@ cisco_nxos_show_license_usage.textfsm, .*, cisco_nxos, sh[[ow]] lic[[ense]] us[[
 cisco_nxos_show_processes_cpu.textfsm, .*, cisco_nxos, sh[[ow]] proc[[esses]] c[[pu]]
 cisco_nxos_show_vrf_interface.textfsm, .*, cisco_nxos, sh[[ow]] vrf int[[erface]]
 cisco_nxos_show_access-lists.textfsm, .*, cisco_nxos, sh[[ow]] acc[[ess-lists]]
-cisco_nxos_show_environments.textfsm, .*, cisco_nxos, sh[[ow]] envi[[ronments]]
```
Comment (FragmentedPacket) [inline templates/index:279]: @diepes After digging into this a bit more during the review, we don't believe `show environments` is a valid command so we should be able to actually remove that and just keep the following line (notice the edit as well).

`cisco_nxos_show_environment.textfsm, .*, cisco_nxos, sh[[ow]] env[[ironment]]`

File: templates/cisco_nxos_show_environment.textfsm [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  ^Power\s+Actual\s+Total -> PowerOut
+  ^Power\s+Actual\s+Actual\s+Total -> PowerOutIn
+  ^.+ -> Error "ERR:PowerSupply"
+
+PowerOut
+  ^${POWER_SUPPLY}\s+${POWER_SUPPLY_MODEL}\s+${POWER_SUPPLY_OUTPUT}\s+\w+\s+${POWER_SUPPLY_CAPACITY}\s+\w+\s+${POWER_SUPPLY_STATUS} -> Record
+  # Filter junk
+  ^Supply\s+Model\s+Output
+  ^\s+\(Watts\s\)\s+
+  ^-----+\s+-+
```
Comment (FragmentedPacket) [inline templates/cisco_nxos_show_environment.textfsm:29]: We could simplify all these.

```suggestion
  ^-+
```

File: templates/cisco_nxos_show_environment.textfsm [hunk 3]
```diff
@@ -0,0 +1,54 @@
+Value Required POWER_SUPPLY (\d+)
```
Comment (FragmentedPacket) [inline templates/cisco_nxos_show_environment.textfsm:1]: We should re-use the other template and then add the additional fixes required for #790.

Also, make sure the values are the same from the old template as well.

**Review by FragmentedPacket — CHANGES_REQUESTED** [review]

File: templates/index (hunk 1 above)
Comment (FragmentedPacket) [reply templates/index:279]: Do you mind updating the new line with the above so users can use `show env` command?

Comment (diepes) [reply templates/index:279]: Fixed

File: templates/cisco_nxos_show_environment.textfsm [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  ^${FAN}\s+\S+\s+\S+\s+\S+\s+${FAN_STATUS}\s*$$ -> Record
+  ^\s*$$ -> Start
+  # Fan Ignore
+  ^Fan\s+Zone\s+Speed
+  ^-+\s*$$
+  ^. -> Error
+
+Temperature
+  ^Module\s+Sensor\s+MajorThresh\s+MinorThres\s+CurTemp\s+Status
+  ^${TEMPERATURE_MODULE}\s+${TEMPERATURE_SENSOR}(\s*\(\S+\))?\s+${TEMPERATURE_MAJOR_THRESH}\s+${TEMPERATURE_MINOR_THRESH}\s+${TEMPERATURE_CURRENT}\s+${TEMPERATURE_STATUS}\s* -> Record
```
Comment (FragmentedPacket) [inline templates/cisco_nxos_show_environment.textfsm:51]: It doesn't appear this line is working the same as the old template as it captured the `(s3)` after crossbar.

File: templates/cisco_nxos_show_environment.textfsm [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+Value FAN_STATUS (\S+)
+Value TEMPERATURE_MODULE (\d+)
+Value TEMPERATURE_SENSOR (\S+)
+#([^ \t\r\n\f])
+Value TEMPERATURE_MAJOR_THRESH (\d+)
+Value TEMPERATURE_MINOR_THRESH (\d+)
+Value TEMPERATURE_CURRENT (\d+)
+Value TEMPERATURE_STATUS (\S+)
+
+Start
```
Comment (FragmentedPacket) [inline templates/cisco_nxos_show_environment.textfsm:18]:
```suggestion
Start
  # Note: 2020-12 this template is broken, mashing 4 tables into one.
  ^Power\s+Supply:
  ^Voltage
  ^Power\s+Actual\s+Total -> Power
  ^Power\s+Actual\s+Actual\s+Total -> Power
  ^Fan:\s*$$ -> Fan
  ^Temperature:\s*$$ -> Temperature
  #^Fan\s+Model\s+Hw\s+(Direction\s+)?Status -> Fan

Power
  # Capture the remaining headers
  ^Supply\s+Model\s+Output\s+Capacity\s+Status
  ^Supply\s+Model\s+Output\s+Input\s+Capacity\s+Status
  ^\s+\(Watts\s\)
  ^-+
  # Start capturing the data for the two different types of outputs
  ^${POWER_SUPPLY}\s+${POWER_SUPPLY_MODEL}\s+${POWER_SUPPLY_OUTPUT}\s+\w+\s+${POWER_SUPPLY_CAPACITY}\s+\w+\s+${POWER_SUPPLY_STATUS} -> Record
  ^${POWER_SUPPLY}\s+${POWER_SUPPLY_MODEL}\s+${POWER_SUPPLY_OUTPUT}\s+\w+\s+${POWER_SUPPLY_INPUT}\s+\w+\s+${POWER_SUPPLY_CAPACITY}\s+\w+\s+${POWER_SUPPLY_STATUS} -> Record
  ^Power\s+Usage\s+Summary: -> Start
  ^\s*$$ -> Start
  ^. -> Error
```

Sorry this template is pretty complicated. Above are some changes I would personally make for readability and transferring states closer to the data that will be captured.

**Review by FragmentedPacket — CHANGES_REQUESTED** [review]

File: templates/index [hunk 6]
```diff
@@ -276,9 +276,9 @@ cisco_nxos_show_license_usage.textfsm, .*, cisco_nxos, sh[[ow]] lic[[ense]] us[[
...
 cisco_nxos_show_ip_adjacency.textfsm, .*, cisco_nxos, sh[[ow]] ip ad[[jacency]]
 cisco_nxos_show_ip_interface.textfsm, .*, cisco_nxos, sh[[ow]] ip int[[erface]]
+cisco_nxos_show_environment.textfsm, .*, cisco_nxos, sh[[ow]] envi[[ronment]]
```
[the remark this answers is not in the document]
Comment (diepes) [reply templates/index]: Done

File: templates/cisco_nxos_show_environment.textfsm [hunk 7]
```diff
@@ -0,0 +1,55 @@
+Value POWER_SUPPLY (\d+)
+Value POWER_SUPPLY_MODEL (\S+)
+Value POWER_SUPPLY_OUTPUT (\d+)
+Value POWER_SUPPLY_INPUT (\d+)
+Value POWER_SUPPLY_CAPACITY (\d+)
+Value POWER_SUPPLY_STATUS (\w+)
+Value POWER_SUPPLY_VOLTAGE (\d+)
```
[the remark this answers is not in the document]
Comment (diepes) [reply templates/cisco_nxos_show_environment.textfsm:7]: Ok

**Review by diepes — COMMENTED** [review]
The only change in the original output now is the addition of Value.
power_supply_input: ""
That is matched in new test template.

This template is still fundamentally broken in mashing together 3 different tables into one. (Power supply's, Fans's, Temperatures)

**Comment by FragmentedPacket** [comment]
@diepes It appears the second `.raw` is not capturing correctly and needs some more template modification. If you look at it, it is not capturing the input and then capturing the wrong values for the others after.

**Comment by diepes** [comment]
> @diepes It appears the second `.raw` is not capturing correctly and needs some more template modification. If you look at it, it is not capturing the input and then capturing the wrong values for the others after.

tests/cisco_nxos/show_environment/cisco_nxos_show_environment2.yml 
Made 1st capture for Power supply values more specific adding line end to prevent it matching values in second 2.raw file.
Also unknown FAN being captured, fixed.

**Comment by FragmentedPacket** [comment]
Did you mean to close this? I'll try and review later tonight if you re-open it.

Thanks again for all your work into these templates.

**Comment by diepes** [comment]
no sorry, i thought it was close the reply/request

**Review by FragmentedPacket — APPROVED** [review]

**Pull request merged**